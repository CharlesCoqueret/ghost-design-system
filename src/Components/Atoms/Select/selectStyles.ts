import alpha from 'color-alpha';
import { StylesConfig } from 'react-select';

import { IOption } from '../../Molecules/SelectField/types';

interface ICustomStylesProps {
  controlErrorColor?: string;
  controlFocusColor?: string;
  fontColor?: string;
  optionSelectedColor?: string;
  optionFocusColor?: string;
  isInError?: boolean;
}

export const customStyles = (props: ICustomStylesProps = {}): StylesConfig<IOption> => {
  const {
    controlErrorColor = 'rgb(255,52,24);',
    controlFocusColor = 'rgb(1,82,129)',
    fontColor = 'rgb(0,0,0)',
    optionFocusColor = 'rgb(228,228,228)',
    optionSelectedColor = 'rgb(38,186,212)',
    isInError = false,
  } = props;

  return {
    control: (provided, state) => {
      let border = provided.border;
      let hover = { border: `1px solid ${controlFocusColor}` };

      if (state.isFocused) {
        border = `1px solid ${controlFocusColor}`;
      }

      if (isInError) {
        border = `1px solid ${controlErrorColor}`;
        hover = { border: `1px solid ${controlErrorColor}` };
      }

      return {
        ...provided,
        outline: 0,
        boxShadow: 'none',
        border,
        ':hover': hover,
      };
    },
    multiValue: (provided) => {
      return { ...provided, minWidth: 'unset', maxWidth: '100%' };
    },
    multiValueLabel: (provided) => {
      return { ...provided, textOverflow: 'unset', fontSize: 'unsert' };
    },
    input: (provided, state) => {
      if (state.isMulti && state.hasValue) {
        return { provided, order: 5, width: 'fit-content' };
      }
      return provided;
    },
    option: (provided, state) => {
      let backgroundColor = provided.backgroundColor;
      let active = provided[':active'];
      const color = fontColor;
      const cursor = 'pointer';

      if (state.isFocused) {
        backgroundColor = alpha(optionFocusColor, 0.33);
        active = { ...active, backgroundColor: alpha(optionFocusColor, 0.66) };
      }

      if (state.isSelected) {
        backgroundColor = alpha(optionSelectedColor, 0.33);
        active = {
          ...active,
          backgroundColor: alpha(optionSelectedColor, 0.66),
        };
      }

      if (state.isFocused && state.isSelected) {
        backgroundColor = alpha(optionSelectedColor, 0.5);
        active = {
          ...active,
          backgroundColor: alpha(optionSelectedColor, 0.5),
        };
      }

      return {
        ...provided,
        ':active': active,
        backgroundColor,
        whiteSpace: 'nowrap',
        textOverflow: 'ellipsis',
        maxWidth: '100%',
        width: '100%',
        overflow: 'hidden',
        cursor,
        color,
      };
    },
    valueContainer: (provided) => {
      return { ...provided, flexWrap: 'nowrap' };
    },
    container: (provided) => {
      return { ...provided, maxWidth: '100%', width: '100%' };
    },
  };
};
