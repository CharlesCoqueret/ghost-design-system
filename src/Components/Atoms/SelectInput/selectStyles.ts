import { StylesConfig } from 'react-select';

import { IOption } from './types';

interface ICustomStylesProps {
  isInError?: boolean;
}

export const customStyles = (props: ICustomStylesProps = {}): StylesConfig<IOption> => {
  const { isInError = false } = props;

  return {
    control: (provided, state) => {
      let border = `1px solid rgb(var(--theme-provider-chalk))`;
      let hover = { border: '1px solid rgb(var(--theme-provider-scooter))' };

      if (state.isFocused) {
        border = '1px solid rgb(var(--theme-provider-scooter))';
      }

      if (isInError) {
        border = '1px solid rgb(var(--theme-provider-error))';
        hover = { border: '1px solid rgb(var(--theme-provider-error))' };
      }

      return {
        ...provided,
        outline: 0,
        boxShadow: 'none',
        minHeight: 'unset',
        border,
        ':hover': hover,
      };
    },
    input: (provided, state) => {
      let multi = {};

      if (state.isMulti && state.hasValue) {
        multi = {
          whiteSpace: 'nowrap',
          textOverflow: 'ellipsis',
          overflow: 'hidden',
        };
      }
      return { ...provided, ...multi };
    },
    option: (provided, state) => {
      let backgroundColor = provided.backgroundColor;
      let active = provided[':active'];
      const color = 'rgb(var(--theme-provider-black))';
      const cursor = 'pointer';

      if (state.isFocused && state.isSelected) {
        backgroundColor = 'rgba(var(--theme-provider-scooter), 0.5)';
        active = {
          ...active,
          backgroundColor: 'rgba(var(--theme-provider-scooter), 0.5)',
        };
      } else if (state.isFocused) {
        backgroundColor = 'rgba(var(--theme-provider-chalk), 0.33)';
        active = { ...active, backgroundColor: 'rgba(var(--theme-provider-chalk), 0.66)' };
      } else if (state.isSelected) {
        backgroundColor = 'rgba(var(--theme-provider-scooter), 0.33)';
        active = {
          ...active,
          backgroundColor: 'rgba(var(--theme-provider-scooter), 0.66)',
        };
      }

      return {
        ...provided,
        ':active': active,
        backgroundColor,
        whiteSpace: 'nowrap',
        textOverflow: 'ellipsis',
        maxWidth: '100%',
        width: '100%',
        overflow: 'hidden',
        cursor,
        color,
      };
    },
    valueContainer: (provided) => {
      return { ...provided, flexWrap: 'nowrap', whiteSpace: 'nowrap', textOverflow: 'ellipsis', padding: '0px 8px' };
    },
    container: (provided) => {
      return { ...provided, maxWidth: '100%', width: '100%' };
    },
    menu: (provided) => {
      return { ...provided, boxShadow: '0px 2px 10px 0px rgba(var(--theme-provider-black), 0.25)' };
    },
  };
};
