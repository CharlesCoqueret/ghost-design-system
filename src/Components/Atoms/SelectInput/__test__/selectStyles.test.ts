import { GroupBase, OptionProps } from 'react-select';
import { customStyles } from '../selectStyles';

describe('customStyles method', () => {
  it('customStyles provides object', async () => {
    const styles = customStyles();

    expect(styles.control).toBeDefined();
    expect(styles.input).toBeDefined();
    expect(styles.option).toBeDefined();
    expect(styles.valueContainer).toBeDefined();
    expect(styles.container).toBeDefined();
  });

  it('customStyles option', async () => {
    const styles = customStyles();

    expect(styles.option).not.toBeUndefined();

    if (styles.option) {
      expect(
        styles.option({ height: '5px' }, { isFocused: true, isSelected: true } as OptionProps<
          { value: string | number; label: string },
          true,
          GroupBase<{ value: string | number; label: string }>
        >),
      )?.toMatchSnapshot();
      expect(
        styles.option({ height: '5px' }, { isSelected: true } as OptionProps<
          { value: string | number; label: string },
          true,
          GroupBase<{ value: string | number; label: string }>
        >),
      )?.toMatchSnapshot();
    }
  });
});
