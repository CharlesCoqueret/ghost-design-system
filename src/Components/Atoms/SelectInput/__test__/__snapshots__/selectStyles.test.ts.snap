// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customStyles method customStyles option 1`] = `
Object {
  ":active": Object {
    "backgroundColor": "rgba(38,186,212,0.5)",
  },
  "backgroundColor": "rgba(38,186,212,0.5)",
  "color": "rgb(0,0,0)",
  "cursor": "pointer",
  "height": "5px",
  "maxWidth": "100%",
  "overflow": "hidden",
  "textOverflow": "ellipsis",
  "whiteSpace": "nowrap",
  "width": "100%",
}
`;

exports[`customStyles method customStyles option 2`] = `
Object {
  ":active": Object {
    "backgroundColor": "rgba(38,186,212,0.66)",
  },
  "backgroundColor": "rgba(38,186,212,0.33)",
  "color": "rgb(0,0,0)",
  "cursor": "pointer",
  "height": "5px",
  "maxWidth": "100%",
  "overflow": "hidden",
  "textOverflow": "ellipsis",
  "whiteSpace": "nowrap",
  "width": "100%",
}
`;
